controller_manager/src/controller_manager.cpp

```diff
@@ -49,56 +50,104 @@ rclcpp::NodeOptions get_cm_node_options()
   rclcpp::NodeOptions node_options;
   // Required for getting types of controllers to be loaded via service call
   node_options.allow_undeclared_parameters(true);
+  node_options.automatically_declare_parameters_from_overrides(true);
   return node_options;
 }
 
 ControllerManager::ControllerManager(
-  std::shared_ptr<hardware_interface::RobotHardware> hw,
   std::shared_ptr<rclcpp::Executor> executor,
   const std::string & manager_node_name)
 : rclcpp::Node(manager_node_name, get_cm_node_options()),
-  hw_(hw),
+  resource_manager_(std::make_shared<hardware_interface::ResourceManager>()),
   executor_(executor),
   loader_(std::make_shared<pluginlib::ClassLoader<controller_interface::ControllerInterface>>(
       kControllerInterfaceName, kControllerInterface))
 {
+  std::string robot_description = "";
+  get_parameter("robot_description", robot_description);
+  if (robot_description.empty()) {
+    throw std::runtime_error("unable to initialize resource manager, no robot description found.");
+  }
+
+  resource_manager_->initialize_from_urdf(robot_description);
+
+  init_services();
+}
+
+ControllerManager::ControllerManager(
+  std::unique_ptr<hardware_interface::ResourceManager> resource_manager,
+  std::shared_ptr<rclcpp::Executor> executor,
+  const std::string & manager_node_name)
+: rclcpp::Node(manager_node_name, get_cm_node_options()),
+  resource_manager_(
+    std::shared_ptr<hardware_interface::ResourceManager>(resource_manager.release())),
```

Comment:
This looks like a hack, since `resource_manager_` is a unique_ptr.

Reply:
Good catch. Addressed in 4372131